WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java

```diff
@@ -395,7 +396,7 @@ public static boolean isPostUploadingOrQueued(PostModel post) {
         }
```

Comment:
>I think the race condition stems from this synchronized call within isPostUploadingOrQueued. I'm not sure if, as an alternative to the local variables, we can synchronize here too? (maybe synchronizing on sInstance can reach deadlock?)

I'm not sure I follow, can you please elaborate? Thanks!

My understanding was that we check `instance == null`, but other thread (eg. main thread) might call `onDestroy` before we reach the point where the current thread uses `instance` -> which can result in NPE, since the Main Thread has just cleared the `instance`.

>One question though, will there be any undefined behavior if we access the service instance after its super.onDestroy() has been called?

Good question! Imo it's still a regular class and moreover, we actually use the instance just to access `UploadStore` so I don't think there'll be any undefined behavior.

Reply:
> I'm not sure I follow, can you please elaborate? Thanks!

In walking the code path for a blocking call on one thread that would generally allow another thread to "see" a different state, I found that synchronized block. Of course a race condition can occur anyway, as execution order on those separate threads is not guaranteed.. but I was looking specifically for something that could "slow one thread down", or be blocking in some way, making the race condition more likely. I was thinking that may be another approach to consider would be to synchronize, rather than keep a local copy, but my hesitation with that "alternative suggestion" of mine is around the potential for dead-lock.

**TL;DR** At first I had a thought we could synchronize to solve this, but have second thoughts because of the other synchronized block.

Btw, I think your approach is simple and effective. I'm mentioning this synchronized approach mainly for completeness.

>  Imo it's still a regular class and moreover, we actually use the instance just to access UploadStore so I don't think there'll be any undefined behavior.

Ok, I thought that might be the case, thanks for confirming! :+1: